src/keygen/output.rs

```diff
@@ -0,0 +1,232 @@
+            error!("Keygen output does not produce a valid public key.");
+            InternalError::InternalInvariantFailed
+        })
+    }
+
+    pub(crate) fn public_key_shares(&self) -> &[KeySharePublic] {
+        &self.public_key_shares
+    }
+
+    pub(crate) fn private_key_share(&self) -> &KeySharePrivate {
+        &self.private_key_share
+    }
+
+    /// This could be made public if appropriate
+    #[cfg(test)]
+    pub(crate) fn rid(&self) -> &[u8; 32] {
+        &self.rid
+    }
+
+    /// Create a new `Output` from its constitutent parts.
+    ///
+    /// This method should only be used with components that were previously
+    /// derived via the [`Output::into_parts()`] method; the calling application
+    /// should not try to form public and private key shares independently.
+    ///
+    /// The provided components must satisfy the following properties:
+    /// - There is a valid key pair -- that is, the public key corresponding to
+    ///   the private key share must be contained in the list of public shares.
+    /// - The public key shares must be from a unique set of participants
+    pub fn from_parts(
+        public_key_shares: Vec<KeySharePublic>,
+        private_key_share: KeySharePrivate,
+        rid: [u8; 32],
+    ) -> Result<Self> {
+        let pids = public_key_shares
+            .iter()
+            .map(KeySharePublic::participant)
+            .collect::<HashSet<_>>();
+        if pids.len() != public_key_shares.len() {
+            error!("Tried to create a keygen output using a set of public material from non-unique participants");
+            Err(CallerError::BadInput)?
+        }
+
+        let expected_public_share = private_key_share.public_share()?;
+        if !public_key_shares
+            .iter()
+            .any(|share| share.as_ref() == &expected_public_share)
+        {
+            error!("Tried to create a keygen output using a private share with no corresponding public share");
+            Err(CallerError::BadInput)?
+        }
+
+        Ok(Self {
+            public_key_shares,
+            private_key_share,
+            rid,
+        })
+    }
+
+    /// Decompose the `Output` into its constituent parts.
+    ///
+    /// # 🔒 Storage requirements
+    /// The [`KeySharePrivate`] must be stored securely by the calling
+    /// application, and a best effort should be made to drop it from memory
+    /// after it's securely stored.
+    ///
+    /// The public components (including the byte array and the public key
+    /// shares) can be stored in the clear.
+    pub fn into_parts(self) -> (Vec<KeySharePublic>, KeySharePrivate, [u8; 32]) {
+        (self.public_key_shares, self.private_key_share, self.rid)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{utils::testing::init_testing, ParticipantConfig, ParticipantIdentifier};
+    use rand::{CryptoRng, Rng, RngCore};
+
+    impl Output {
```

Comment:
These two simulate methods were copied from the other file. The only update is that they now use the `from_parts` method to create `Output` instead of manually constructing it.